pkgs/cli_config/lib/src/config.dart

```diff
@@ -0,0 +1,387 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'cli_parser.dart';
+import 'cli_provider.dart';
+import 'environment_parser.dart';
+import 'environment_provider.dart';
+import 'file_parser.dart';
+import 'file_provider.dart';
+import 'provider.dart';
+
+/// A hierarchical configuration object.
+///
+/// Configuration can be provided as commandline arguments, environment
+/// variables and configuration files. This configuration object makes
+/// these accessible with a uniform API.
+///
+/// Configuration can be provided in three ways:
+/// 1. commandline argument defines as `-Dsome_key=some_value`,
+/// 2. environment variables as `SOME_KEY=some_value`, and
+/// 3. config files as JSON or YAML as `{'some_key': 'some_value'}`.
+///
+/// The default lookup behavior is that commandline argument defines take
```

Comment:
What terminology should we use for the different config sources? config sources? config providers?

Reply:
Config sources is understandable enough.

Reply:
I'll probably rename all the `Provider` to `Source` in the codebase and `provider` to `source` in doc-comments.